### PR TITLE
planner: fix wrong prepare plan after isolation read changed

### LINKED ISSUE #16256: plan cache: the query plan doesn't change when tidb_isolation_read_engines be set to tifalsh
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?

```
use test;
drop table if exists t;
create table t(a int);
select connection_id();
ALTER TABLE t SET TIFLASH REPLICA 1;
do sleep(20);
set @@session.tidb_isolation_read_engines='tikv';
prepare stmt from "select * from t";
execute stmt;
# explain for connection id;
set @@session.tidb_isolation_read_engines='tiflash';
execute stmt;
# explain for connection id;
```

### 2. What did you expect to see?
The last sql's plan should use tiflash
mysql> explain for connection 1;
+-----------------------+----------+-----------+---------------+--------------------------------+
| id                    | estRows  | task      | access object | operator info                  |
+-----------------------+----------+-----------+---------------+--------------------------------+
| TableReader_5         | 10000.00 | root      |               | data:TableFullScan_4           |
| └─TableFullScan_4 | 10000.00 | cop[**tiflash**] | table:t       | keep order:false, stats:pseudo |
+-----------------------+----------+-----------+---------------+--------------------------------+
2 rows in set (0.04 sec)

### 3. What did you see instead?
mysql> explain for connection 1;
+-----------------------+----------+-----------+---------------+--------------------------------+
| id                    | estRows  | task      | access object | operator info                  |
+-----------------------+----------+-----------+---------------+--------------------------------+
| TableReader_5         | 10000.00 | root      |               | data:TableFullScan_4           |
| └─TableFullScan_4 | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo |
+-----------------------+----------+-----------+---------------+--------------------------------+
2 rows in set (0.04 sec)

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
```
Release Version: v4.0.0-beta.2-151-g78922a8
Git Commit Hash: 78922a8abef2f21e4d8990d37c6c67ac7b0eef42
Git Branch: HEAD
UTC Build Time: 2020-04-07 05:45:12
GoVersion: go1.13.7
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve?

Issue Number: close #16256

Problem Summary: After isolation read changed, the prepared statement should be re-prepared.

### What is changed and how it works?

What's Changed: clear prepared plan when isolation read changes.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU

### Release note

- Make plan cache consider isolation read engines.

### CODE REVIEW

File: planner/core/common_plans.go [hunk 1]
```diff
@@ -229,20 +230,24 @@ func (e *Execute) OptimizePreparedPlan(ctx context.Context, sctx sessionctx.Cont
 		}
 	}
 
-	if prepared.SchemaVersion != is.SchemaMetaVersion() {
+	if prepared.SchemaVersion != is.SchemaMetaVersion() || !reflect.DeepEqual(preparedObj.IsolationEngines, e.ctx.GetSessionVars().IsolationReadEngines) {
```
Comment (crazycs520) [inline planner/core/common_plans.go:233]: How about not use `reflect.DeepEqual`?

```go
func prepare() (map[string]struct{}, map[string]struct{}) {
	m1 := make(map[string]struct{}, 3)
	m1["tikv"] = struct{}{}
	m1["tidb"] = struct{}{}
	m1["tiflash"] = struct{}{}
	m2 := make(map[string]struct{}, 3)
	m2["tikv"] = struct{}{}
	m2["tidb"] = struct{}{}
	m2["tiflash"] = struct{}{}
	return m1, m2
}

func Benchmark_cmp1(b *testing.B) {
	m1, m2 := prepare()
	for i := 0; i < b.N; i++ {
		reflect.DeepEqual(m1, m2)
	}
}

func Benchmark_cmp2(b *testing.B) {
	m1, m2 := prepare()
	for i := 0; i < b.N; i++ {
		checkEqualMap(m1, m2)
	}
}

func checkEqualMap(m1, m2 map[string]struct{}) bool {
	if len(m1) != len(m2) {
		return false
	}
	for k := range m1 {
		if _, ok := m2[k]; !ok {
			return false
		}
	}
	return true
}
```

```go
▶ go test -run '^$' -bench .
goos: darwin
goarch: amd64
pkg: github.com/crazycs520/test
Benchmark_cmp1-12        2272569               521 ns/op
Benchmark_cmp2-12       21689436                54.1 ns/op
PASS
ok      github.com/crazycs520/test      2.955s

```

**Comment by lawyerphx** [comment]
LGTM

File: executor/prepared_test.go [hunk 2]
```diff
@@ -14,7 +14,12 @@
 package executor_test
 
 import (
+	"fmt"
```
Comment (zz-jason) [inline executor/prepared_test.go:17]: please add an empty line.

**Comment by zz-jason** [comment]
> clear prepared plan when isolation read changes.

I prefer to add isolation read engines to the key of the prepared plan cache.

@eurekaka What's your opinion?

**Comment by eurekaka** [comment]
> > clear prepared plan when isolation read changes.
> 
> I prefer to add isolation read engines to the key of the prepared plan cache. It's different from schema change. schema version change is unpreventable, but isolation read engine change is. Once the user changed back to the original value of the isolation read engine, the cache can be used again at that time.
> 
> @eurekaka What's your opinion?

+1

**Comment by lzmhhh123** [comment]
> > > clear prepared plan when isolation read changes.
> > 
> > 
> > I prefer to add isolation read engines to the key of the prepared plan cache. It's different from schema change. schema version change is unpreventable, but isolation read engine change is. Once the user changed back to the original value of the isolation read engine, the cache can be used again at that time.
> > @eurekaka What's your opinion?
> 
> +1

OK, I'll change it to that way.

File: executor/prepared_test.go [hunk 3]
```diff
@@ -53,3 +59,41 @@ func (s *testSuite1) TestIgnorePlanCache(c *C) {
...
+	rows := tk.MustQuery(fmt.Sprintf("explain for connection %d", tkProcess.ID)).Rows()
+	c.Assert(rows[len(rows)-1][2], Equals, "cop[tikv]")
+
+	tk.MustExec("set @@session.tidb_isolation_read_engines='tiflash'")
+	tk.MustExec("execute stmt")
+	tkProcess = tk.Se.ShowProcess()
+	ps = []*util.ProcessInfo{tkProcess}
+	tk.Se.SetSessionManager(&mockSessionManager1{PS: ps})
+	rows = tk.MustQuery(fmt.Sprintf("explain for connection %d", tkProcess.ID)).Rows()
+	c.Assert(rows[len(rows)-1][2], Equals, "cop[tiflash]")
```
Comment (zz-jason) [inline executor/prepared_test.go:98]: it's better to check the plan cache content as well.

**Pull request merged**